hydra-node/src/Hydra/Chain/Direct/State.hs

```diff
@@ -0,0 +1,400 @@
+-- wallet UTxO.
+
+-- | Initialize a head from an 'StIdle' state. This does not change the state
+-- but produces a transaction which, if observed, does apply the transition.
+initialize ::
+  HeadParameters ->
+  [VerificationKey PaymentKey] ->
+  TxIn ->
+  OnChainHeadState 'StIdle ->
+  Tx
+initialize params cardanoKeys seedInput OnChainHeadState{networkId} = do
+  initTx networkId cardanoKeys params seedInput
+
+commit ::
+  UTxO ->
+  OnChainHeadState 'StInitialized ->
+  Either (PostTxError Tx) Tx
+commit utxo st@OnChainHeadState{networkId, ownParty, ownVerificationKey, stateMachine} = do
+  case ownInitial ownVerificationKey initialInitials of
+    Nothing ->
+      Left (CannotFindOwnInitial{knownUTxO = getKnownUTxO st})
+    Just initial ->
+      case UTxO.pairs utxo of
+        [aUTxO] -> do
+          Right $ commitTx networkId ownParty (Just aUTxO) initial
+        [] -> do
+          Right $ commitTx networkId ownParty Nothing initial
+        _ ->
+          Left (MoreThanOneUTxOCommitted @Tx)
+ where
+  Initialized
+    { initialInitials
+    } = stateMachine
+
+abort ::
+  HasCallStack =>
+  OnChainHeadState 'StInitialized ->
+  Tx
+abort OnChainHeadState{networkId, stateMachine} = do
+  let (i, _, dat, _) = initialThreadOutput
+      initials = Map.fromList $ map drop2nd initialInitials
+      commits = Map.fromList $ map tripleToPair initialCommits
+   in case abortTx networkId (i, dat) initials commits of
+        Left err ->
+          -- FIXME: Exception with MonadThrow?
+          error $ show err
+        Right tx ->
+          tx
+ where
+  Initialized
+    { initialThreadOutput
+    , initialInitials
+    , initialCommits
+    } = stateMachine
+
+collect ::
+  OnChainHeadState 'StInitialized ->
+  Tx
+collect OnChainHeadState{networkId, stateMachine} = do
+  let (i, _, dat, parties) = initialThreadOutput
+      commits = Map.fromList $ fmap tripleToPair initialCommits
+   in collectComTx networkId (i, dat, parties) commits
+ where
+  Initialized
+    { initialThreadOutput
+    , initialCommits
+    } = stateMachine
+
+close ::
+  ConfirmedSnapshot Tx ->
+  OnChainHeadState 'StOpen ->
+  Tx
+close confirmedSnapshot OnChainHeadState{stateMachine} = do
+  let (sn, sigs) =
+        case confirmedSnapshot of
+          ConfirmedSnapshot{snapshot, signatures} -> (snapshot, signatures)
+          InitialSnapshot{snapshot} -> (snapshot, mempty)
+      (i, o, dat, _) = openThreadOutput
+   in closeTx sn sigs (i, o, dat)
+ where
+  Open{openThreadOutput} = stateMachine
+
+fanout ::
+  UTxO ->
+  OnChainHeadState 'StClosed ->
+  Tx
+fanout utxo OnChainHeadState{stateMachine} = do
+  let (i, _, dat, _) = closedThreadOutput
+   in fanoutTx utxo (i, dat)
+ where
+  Closed{closedThreadOutput} = stateMachine
+
+-- Observing Transitions
+
+class HasTransition st where
+  observeTx ::
+    Tx ->
+    OnChainHeadState st ->
+    Maybe (OnChainTx Tx, SomeOnChainHeadState)
```

Comment:
Do we even require a type class after all? `OnChainHeadState st` is still a single type and wouldn't we be able to implement `observeTx :: Tx -> OnChainHeadState st -> Maybe (OnChainTx Tx, SomeOnChainHeadState)` with something like the following?

```hs
observeTx tx OnChainHeadState{networkId, ownParty, stateMachine} =
  case stateMachine of
    Idle{} ->
      (event, observation) <- observeInitTx networkId ownParty tx
      -- etc
    Initialized{initialInitials, initialCommits} ->
      observeCommit <|> observeCollectCom <|> observeAbort
      -- etc
-- ... etc using all the same code in the individual instances
```

Reply:
Yes and no, I had the type-class originally because the instances were fully capturing the transitions (see #228). But, this may be _too much dependent type_. 

Still, having a type-class here allows for overloading `observeTx` in its arguments. Without that, we would need to pattern-match on the call-site to figure out in which state we are, and call the right transition function, and put the state back in an existential. That's perhaps doable (and actually, a good in-between proposition between the current state of affairs and #228 :sweat_smile: ...)